sycl/plugins/rocm/pi_rocm.cpp

```diff
@@ -0,0 +1,4617 @@
+  }
+  default:
+    __SYCL_PI_HANDLE_UNKNOWN_PARAM_NAME(param_name);
+  }
+  cl::sycl::detail::pi::die("Platform info request not implemented");
+  return {};
+}
+
+/// \param devices List of devices available on the system
+/// \param num_devices Number of elements in the list of devices
+/// Requesting a non-GPU device triggers an error, all PI HIP devices
+/// are GPUs.
+///
+pi_result rocm_piDevicesGet(pi_platform platform, pi_device_type device_type,
+                            pi_uint32 num_entries, pi_device *devices,
+                            pi_uint32 *num_devices) {
+
+  pi_result err = PI_SUCCESS;
+  const bool askingForDefault = device_type == PI_DEVICE_TYPE_DEFAULT;
+  const bool askingForGPU = device_type & PI_DEVICE_TYPE_GPU;
+  const bool returnDevices = askingForDefault || askingForGPU;
+
+  size_t numDevices = returnDevices ? platform->devices_.size() : 0;
+
+  try {
+    if (num_devices) {
+      *num_devices = numDevices;
+    }
+
+    if (returnDevices && devices) {
+      for (size_t i = 0; i < std::min(size_t(num_entries), numDevices); ++i) {
+        devices[i] = platform->devices_[i].get();
+      }
+    }
+
+    return err;
+  } catch (pi_result err) {
+    return err;
+  } catch (...) {
+    return PI_OUT_OF_RESOURCES;
+  }
+}
+
+/// \return PI_SUCCESS if the function is exehipted successfully
+/// HIP devices are always root devices so retain always returns success.
+pi_result rocm_piDeviceRetain(pi_device device) { return PI_SUCCESS; }
+
+pi_result rocm_piContextGetInfo(pi_context context, pi_context_info param_name,
+                                size_t param_value_size, void *param_value,
+                                size_t *param_value_size_ret) {
+
+  switch (param_name) {
+  case PI_CONTEXT_INFO_NUM_DEVICES:
+    return getInfo(param_value_size, param_value, param_value_size_ret, 1);
+  case PI_CONTEXT_INFO_DEVICES:
+    return getInfo(param_value_size, param_value, param_value_size_ret,
+                   context->get_device());
+  case PI_CONTEXT_INFO_REFERENCE_COUNT:
+    return getInfo(param_value_size, param_value, param_value_size_ret,
+                   context->get_reference_count());
+  default:
+    __SYCL_PI_HANDLE_UNKNOWN_PARAM_NAME(param_name);
+  }
+
+  return PI_OUT_OF_RESOURCES;
+}
+
+pi_result rocm_piContextRetain(pi_context context) {
+  assert(context != nullptr);
+  assert(context->get_reference_count() > 0);
+
+  context->increment_reference_count();
+  return PI_SUCCESS;
+}
+
+pi_result rocm_piextContextSetExtendedDeleter(
+    pi_context context, pi_context_extended_deleter function, void *user_data) {
+  context->set_extended_deleter(function, user_data);
+  return PI_SUCCESS;
+}
+
+/// Not applicable to HIP, devices cannot be partitioned.
+///
+pi_result rocm_piDevicePartition(
+    pi_device device,
+    const cl_device_partition_property *properties, // TODO: untie from OpenCL
+    pi_uint32 num_devices, pi_device *out_devices, pi_uint32 *out_num_devices) {
+  return {};
```

Comment:
I'd return something like `PI_INVALID_OPERATION` or `PI_INVALID_DEVICE`.
```suggestion
  return PI_INVALID_OPERATION;
```